cmd/dra-example-kubeletplugin/state.go

```diff
@@ -173,3 +213,97 @@ func (s *DeviceState) prepareDevices(claim *resourceapi.ResourceClaim) (Prepared
 func (s *DeviceState) unprepareDevices(claimUID string, devices PreparedDevices) error {
 	return nil
 }
+
+func (p *PreparedDevice) applyConfig() error {
+	return nil
+}
+
+// GetDevices extracts the list of drapbv1.Devices from PreparedDevices.
+func (pds PreparedDevices) GetDevices() []*drapbv1.Device {
+	var devices []*drapbv1.Device
+	for _, pd := range pds {
+		devices = append(devices, &pd.Device)
+	}
+	return devices
+}
+
+// ptrImplementsRuntimeObject is a helper interface for GetOpaqueDeviceConfig.
+//
+// It enforces that a pointer to the type parameter passed to
+// GetOpaqueDeviceConfig implements the runtime.Object interface.
+type ptrImplementsRuntimeObject[T any] interface {
+	*T
+	runtime.Object
+}
+
+// GetOpaqueDeviceConfig returns the config with the highest precedence specified for a device request in a resource claim.
+//
+// Configs can either come from the resource claim itself or from the device
+// class associated with the request. Configs coming directly from the resource
+// claim take precedence over configs coming from the device class. Moreover,
+// configs found earlier in the list of configs attached to its corresponding
+// source take precedence over configs found later in the list for that source.
+//
+// Only the config with the highest precedence in the order specified above
+// will be returned. All other configs will be ignored. If no config is found,
+// the default config passed in will be returned.
+func GetOpaqueDeviceConfig[T any, TI ptrImplementsRuntimeObject[T]](
+	decoder runtime.Decoder,
+	driverName,
+	request string,
+	possibleConfigs []resourceapi.DeviceAllocationConfiguration,
+	defaultConfig T,
+) (T, error) {
+	// Collect all configs in order of reverse precedence.
+	var classConfigs []resourceapi.DeviceConfiguration
+	var claimConfigs []resourceapi.DeviceConfiguration
+	var allConfigs []resourceapi.DeviceConfiguration
+	for _, config := range possibleConfigs {
+		if !slices.Contains(config.Requests, request) {
+			continue
+		}
+		switch config.Source {
+		case resourceapi.AllocationConfigSourceClass:
+			classConfigs = append(classConfigs, config.DeviceConfiguration)
+		case resourceapi.AllocationConfigSourceClaim:
+			claimConfigs = append(claimConfigs, config.DeviceConfiguration)
+		default:
+			return *new(T), fmt.Errorf("invalid config source: %v", config.Source)
+		}
+	}
+	allConfigs = append(allConfigs, classConfigs...)
+	allConfigs = append(allConfigs, claimConfigs...)
+
+	// Walk all configs from the allConfigs slice in the order they were added.
+	// The latest config matching the GroupVersionKind of the defaultConfig
+	// will be returned. If no configs match, then the original defaultConfig
+	// is returned.
+	candidateConfig := defaultConfig
+	for _, config := range allConfigs {
+		if config.Opaque == nil {
+			continue
```

Comment:
I think this should be an error. The semantic of a one-of is that "empty = some future extension". In this case here, the user is asking for something that the driver cannot do. It shouldn't silently continue because the driver needs to be updated.

Reply:
But we don't even know if this config is meant for this driver yet. We only know that later on when we check `if config.Opaque.Driver != driverName {` below.

Reply:
I take it back -- we pre-filter by the request that we know is meant for this driver, so actually both this check and the check below should throw an error if they are not what is expected.

Reply:
Encountering configs meant for other drivers is not an error. The wording of the API is:

https://github.com/kubernetes/kubernetes/blob/95b3fe9f15cdcaf98098be398478e70365b12dd7/pkg/apis/resource/types.go#L583-L584

"Driver is used to determine which kubelet plugin needs to be passed these configuration parameters."

This needs to be updated, kubelet no longer "passes" the configuration. But the intent is the same: it allows requesting a device that could come from driver A or driver B and providing configurations for both. That wouldn't work if driver A errors out on the config for driver B and vice versa.

Reply:
That's fair. I wrote this before you reminded me that we can have an empty list for `requests`.  Otherwise, I would say it should error if the passed in request was not meant for this driver.

Reply:
Updated.

Reply:
> Otherwise, I would say it should error if the passed in request was not meant for this driver.

Same situation: the request might have been defined such that it could be satisfied by two different drivers.

Reply:
True. And then you provide a different config for each driver you know might handle it (knowing only one of them will actually have picked it up). In any case -- its fixed now.